manticore/core/manticore.py

```diff
@@ -523,30 +523,48 @@ def setstate(x, y):
 
         self._publish("will_fork_state", state, expression, solutions, policy)
 
-        # Build and enqueue a state for each solution
+        # Create new states
         children = []
-        for new_value in solutions:
-            with state as new_state:
-                new_state.constrain(expression == new_value)
+        if len(solutions) == 1:
+            # If only one solution don't copy the state but update the current
+            # state instead
+            # Add state constraint
+            new_value = solutions[0]
+            state.constrain(expression == new_value)
+            setstate(state, new_value)
+
+            # Put the state back in ready list
+            self._put_state(state)  # Doesn't change state id
+            self._publish("did_fork_state", state, expression, solutions, policy, [])
```

Comment:
This could be controversial (we're not forking, we're just concretizing data within the current state) but I think it's the most appropriate thing to do here. We don't have a special event for concretizing without forking. 

One thing devs of downstream tools might need to be aware of is that code that consumes the introspection API (like MUI) might need to be tweaked. The state introspection API consumes this event and tracks which states fork from which other states, so with this tweak it'll appear that we're getting multiple fork events that produce the same state.

Reply:
When publishing the fork event the `children` parameter is set to `[]` to indicate that this fork doesn't result in new states being created. 

I agree that this is not really a fork since there is only one solution but in most cases the users would probably still want their callbacks to be executed.